## Урок 6. Взаимосвязь величин. Параметрические и непараметрические показатели корреляции. Корреляционный анализ.

### 1. Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):

zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],

ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].

Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy

Полученные значения должны быть равны.

Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,

а затем с использованием функций из библиотек numpy и pandas.

In [63]:
import numpy as np
import pandas as pd
from math import sqrt

In [64]:
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]

In [65]:
zp_mean = sum(zp) / len(zp)
zp_mean

101.4

In [66]:
n = len(zp)
s = 0
for i in zp:
    s += (i - zp_mean)**2   
zp_sigma = (s / (n-1))**0.5  ## (n - 1) формула для выборки
print(f'Среднее квадратичное отклонение = {zp_sigma}')

Среднее квадратичное отклонение = 62.31318747531162


In [67]:
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]

In [68]:
ks_mean = sum(ks) / len(zp)
ks_mean

709.9

In [69]:
n = len(ks)
s = 0
for i in ks:
    s += (i - ks_mean)**2   
ks_sigma = (s / (n-1))**0.5  ## (n - 1) формула для выборки
print(f'Среднее квадратичное отклонение = {ks_sigma}')

Среднее квадратичное отклонение = 183.99544076477068


In [70]:
sum_ks_zp = 0
for i in range(10):
    sum_ks_zp += (zp[i] - zp_mean) * (ks[i] - ks_mean)
cov_z_k = (sum_ks_zp) / (n - 1)
cov_z_k

10175.377777777778

In [71]:
cov_zp_ks = np.cov(zp_ks)
cov_zp_ks

array([[ 3882.93333333, 10175.37777778],
       [10175.37777778, 33854.32222222]])

$$r_{XY} = \frac{cov_{XY}}{\sigma_{X}\sigma_{Y}}$$

In [72]:
r_zp_ks = cov_z_k / (zp_sigma * ks_sigma)
r_zp_ks

0.8874900920739162

In [73]:
zp_ks = np.array([[35, 45, 190, 200, 40, 70, 54, 150, 120, 110], [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]])
zp_ks

array([[ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110],
       [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]])

In [74]:
coef_zp_ks = np.corrcoef(zp_ks)
coef_zp_ks

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

**Ответ:** ковариация = 10175.38, коэф.корреляции = 0.887

### 2 . Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [75]:
data_iq = [131, 125, 115, 122, 131, 115, 107, 99, 125, 111]

In [76]:
data_iq_mean = sum(data_iq) / len(data_iq)
data_iq_mean

118.1

In [77]:
n = len(data_iq)
s = 0
for i in data_iq:
    s += (i - data_iq_mean)**2   
sigma = (s / (n-1))**0.5  ## (n - 1) формула для выборки
print(f'Среднее квадратичное отклонение = {sigma}')

Среднее квадратичное отклонение = 10.54566788359614


In [78]:
se = sigma / sqrt(n)
se

3.3348329959851224

In [79]:
t_students = 2.2622

In [80]:
se * t_students

7.544059203517544

In [81]:
iq_1 = data_iq_mean + se * t_students
iq_1

125.64405920351754

In [82]:
iq_2 = data_iq_mean - se*t_students
iq_2

110.55594079648245

**Ответ:** [110.6, 125.6]

### 3. Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для математического
ожидания с надежностью 0.95.

In [87]:
D = 25
n = 27
height_mean = 174.2
z_table=1.645

In [88]:
se = sqrt(D) / sqrt(n)
se

0.9622504486493763

In [89]:
print(f'height_1 = {height_mean - se * z_table}')


height_1 = 172.61709801197176


In [90]:
print(f'height_2 = {height_mean + se * z_table}')

height_2 = 175.78290198802821


**Ответ:** [172.62, 175.78]